---
title: "PCA handling"
author: "Ann Von Holle"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 2
editor_options: 
  chunk_output_type: console
---


```{r, echo=F, warning=FALSE, message=FALSE, results='hide'}
library(knitr)
library(ggplot2)
library(data.table)
library(tidyverse)
```

```{r, setup, include=FALSE, cache=FALSE}
# https://bookdown.org/yihui/rmarkdown-cookbook/eng-bash.html
opts_chunk$set(echo=T, cache=F, message=F, warning=F, error=F,
               engine.opts = list(bash = "-l"))

```

# Population stratification work

**Note:** I used code from the following paper to create and examine the PCs to be used for population stratification.

<!-- Sources: 
https://meyer-lab-cshl.github.io/plinkQC/articles/Genomes1000.html
https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html
-->


## Step 1. Run plink to get pruned files to make files 


```{sh, eval=T}
#!/bin/bash

#SBATCH --partition=highmem
#SBATCH --cpus-per-task=15
#SBATCH --mail-user=ann.vonholle@nih.gov
#SBATCH --mail-type=END, FAIL

## Will subset the combo_10 ped file to only the AIMS snps.
## Source of AIMs.txt file is March 2022 email from Katie

## first have to convert old carriage return in AIMs.txt file and replace with space
## https://unix.stackexchange.com/questions/32001/what-is-m-and-how-do-i-get-rid-of-it
## https://support.microfocus.com/kb/doc.php?id=7014821

sed -e "s/\r/ /g" ~/projects/carot_data/AIMs.txt > ~/projects/carot_data/AIMs2.txt

## subsis1c8 from clean.Rmd

plink --bfile ~/projects/carot_data/dat/subsis1c8 \
      --extract ~/projects/carot_data/AIMs2.txt \
      --make-bed \
      --out ~/projects/carot_data/dat/sample_justaims
mv  ~/projects/carot_data/dat/sample_justaims.log ~/projects/carot_data/dat/plink_log/sample_justaims.log

plink --bfile ~/projects/carot_data/1000G/all_phase3 \
      --extract ~/projects/carot_data/AIMs2.txt \
      --allow-extra-chr \
      --make-bed \
      --out ~/projects/carot_data/1000G/all_phase3aims
mv  ~/projects/carot_data/1000G/all_phase3aims.log ~/projects/carot_data/dat/plink_log/all_phase3aims.log

```


```{sh, eval=T}
#!/bin/bash

#SBATCH --partition=highmem
#SBATCH --cpus-per-task=15
#SBATCH --mail-user=ann.vonholle@nih.gov
#SBATCH --mail-type=END, FAIL

qcdir=~/projects/carot_data/dat
refdir=~/projects/carot_data/1000G
name='sample_justaims'
refname='all_phase3aims'

## dir already created.
## mkdir $qcdir/plink_log

## Source: https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html

## from checka.sh

## Filter reference and study data for non A-T or G-C SNPs
## Source: https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#filter-reference-and-study-data-for-non-a-t-or-g-c-snps

awk 'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" \
                        || $5$6 == "AT" || $5$6 == "TA")  {print $2}' \
    $qcdir/$name.bim  > \
    $qcdir/$name.ac_gt_snps

awk 'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" \
                        || $5$6 == "AT" || $5$6 == "TA")  {print $2}' \
    $refdir/$refname.bim  > \
    $qcdir/$refname.ac_gt_snps
   
plink --bfile  $refdir/$refname \
      --exclude $qcdir/$refname.ac_gt_snps \
      --make-bed \
      --out $qcdir/$refname.no_ac_gt_snps
mv  $qcdir/$refname.no_ac_gt_snps.log $qcdir/plink_log/$refname.no_ac_gt_snps.log

plink --bfile  $qcdir/$name \
      --exclude $qcdir/$name.ac_gt_snps \
      --make-bed \
      --out $qcdir/$name.no_ac_gt_snps
mv  $qcdir/$name.no_ac_gt_snps.log $qcdir/plink_log/$name.no_ac_gt_snps.log

## From checka2.sh

## Prune study data
## ==========================================================
## Source: https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#prune-study-data
## Source: https://github.com/meyer-lab-cshl/plinkQC/blob/master/inst/extdata/high-LD-regions-hg19-GRCh37.txt

##      --exclude range  $refdir/$highld \


plink --bfile  $qcdir/$name.no_ac_gt_snps \
      --indep-pairwise 50 5 0.2 \
      --out $qcdir/$name.no_ac_gt_snps
mv  $qcdir/$name.prune.log $qcdir/plink_log/$name.prune.log

plink --bfile  $qcdir/$name.no_ac_gt_snps \
      --extract $qcdir/$name.no_ac_gt_snps.prune.in \
      --make-bed \
      --out $qcdir/$name.pruned
mv  $qcdir/$name.pruned.log $qcdir/plink_log/$name.pruned.log


## Filter reference data for the same SNP set as in study
## ==========================================================
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#filter-reference-data-for-the-same-snp-set-as-in-study

plink --bfile  $refdir/$refname \
      --extract $qcdir/$name.no_ac_gt_snps.prune.in \
      --make-bed \
      --out $qcdir/$refname.pruned
mv  $qcdir/$refname.pruned.log $qcdir/plink_log/$refname.pruned.log


## From checka3.sh
## Source: https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html

## Check and correct chromosome mismatch
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#check-and-correct-chromosome-mismatch
## ==============================================================

awk 'BEGIN {OFS="\t"} FNR==NR {a[$2]=$1; next} \
    ($2 in a && a[$2] != $1)  {print a[$2],$2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim | \
    sed -n '/^[XY]/!p' > $qcdir/$refname.toUpdateChr

plink --bfile $qcdir/$refname.pruned \
      --update-chr $qcdir/$refname.toUpdateChr 1 2 \
      --make-bed \
      --out $qcdir/$refname.updateChr
mv $qcdir/$refname.updateChr.log $qcdir/plink_log/$refname.updateChr.log


## Position mismatch
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#position-mismatch
## ==============================================================

awk 'BEGIN {OFS="\t"} FNR==NR {a[$2]=$4; next} \
    ($2 in a && a[$2] != $4)  {print a[$2],$2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim > \
    $qcdir/${refname}.toUpdatePos
    
    
## Possible allele flips
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#possible-allele-flips
## ==============================================================

awk 'BEGIN {OFS="\t"} FNR==NR {a[$1$2$4]=$5$6; next} \
    ($1$2$4 in a && a[$1$2$4] != $5$6 && a[$1$2$4] != $6$5)  {print $2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim > \
    $qcdir/$refname.toFlip
    
## Upate positions and flip alleles
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#upate-positions-and-flip-alleles
## ==============================================================

plink --bfile $qcdir/$refname.updateChr \
      --update-map $qcdir/$refname.toUpdatePos 1 2 \
      --flip $qcdir/$refname.toFlip \
      --make-bed \
      --out $qcdir/$refname.flipped
mv $qcdir/$refname.flipped.log $qcdir/plink_log/$refname.flipped.log


## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#remove-mismatches
## Remove mismatches
## ==============================================================

awk 'BEGIN {OFS="\t"} FNR==NR {a[$1$2$4]=$5$6; next} \
    ($1$2$4 in a && a[$1$2$4] != $5$6 && a[$1$2$4] != $6$5) {print $2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.flipped.bim > \
    $qcdir/$refname.mismatch

plink --bfile $qcdir/$refname.flipped \
      --exclude $qcdir/$refname.mismatch \
      --make-bed \
      --out $qcdir/$refname.clean
mv $qcdir/$refname.clean.log $qcdir/plink_log/$refname.clean.log



## Merge study genotypes and reference data
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#merge-study-genotypes-and-reference-data
## ==============================================================

plink --bfile $qcdir/$name.pruned  \
      --bmerge $qcdir/$refname.clean.bed $qcdir/$refname.clean.bim \
         $qcdir/$refname.clean.fam  \
      --make-bed \
      --out $qcdir/$name.merge.$refname
mv $qcdir/$name.merge.$refname.log $qcdir/plink_log/$name.merge.$refname.log

## PCA on the merged data
## https://meyer-lab-cshl.github.io/plinkQC/articles/AncestryCheck.html#pca-on-the-merged-data
## ==============================================================

plink --bfile $qcdir/$name.merge.$refname \
      --pca \
      --out $qcdir/$name.$refname
mv $qcdir/$name.$refname.log $qcdir/plink_log/$name.$refname.log


## PCA just on the sample for GWAS
plink --bfile $qcdir/$name.pruned  \
      --pca \
      --out $qcdir/$name.pruned.pca
mv $qcdir/$name.pruned.pca.log $qcdir/plink_log/$name.pruned.pca.log

```


